### ISSUE TITLE
open file(gvwstate.dat.tmp) failed(Permission denied)

### ISSUE THREAD

**Issue by nirbhayc** [codership/galera#345]
Originally reported on MariaDB JIRA [MDEV-8758](https://mariadb.atlassian.net/browse/MDEV-8758) ::

```
Im am install garb from you repositary on Debian jessi, and configure.
#cat /etc/default/garb
GALERA_NODES="192.168.0.1:4567 192.168.0.2:4567 192.168.0.3:4567"
GALERA_GROUP="cluster"
GALERA_OPTIONS="gmcast.listen_addr=192.168.0.4"
LOG_FILE="/var/log/garb.log"
After run i am see in garb.log Warning: open file(gvwstate.dat.tmp) failed(Permission denied)
I am strace garbd and see garb open this file without any path.
I am test set set in start-stop-daemon set option '-d /tmp' no result.
Only set -o root -c root that helped, garb create file in '/' direcory and start as root, but this insecure.
Need parameter for set garb directory, or if this temp file need garb create this file in system /tmp directory.
```

**Comment by philip-galera** [comment]
Duplicate of https://github.com/codership/galera/issues/313

**Comment by aerialls** [comment]
Any update on this issue? garbd won't start after the latest upgrade because of this issue.

Glad to help if I can :)

**Comment by philip-galera** [comment]
You may try to edit the garb init script to put a `cd` command to change the directory to a suitable one before the garbd binary is started.

**Comment by hal58th** [comment]
@philip-galera Someone has tried your solution with no real success in another forum. Any update on this?

**Comment by petzah** [comment]
not sure if it would be correct fix but see 
https://github.com/petzah/galera/commit/2df04fcfc538f79be9bef7ac56e3c61b49c4334f

**Comment by cjbirk** [comment]
@petzah Thanks Petzah.  This seems to make sense.. Let the user configure where the gvwstate.dat gets stored.

**Comment by aramhakobyan** [comment]
and still  after an year same behavior :(

**Comment by LudovicEnault** [comment]
Hi there,
I had the same issue (+ logfile).
To solve it, I overloaded the garb.service configuration (systemd) and hacking 
/etc/systemd/system/garb.service.d/01-workingdirectory.conf:
```
[Service]
PermissionsStartOnly=true
ExecStartPre=/usr/bin/mkdir -p            /var/lib/garb
ExecStartPre=/usr/bin/chown nobody:nobody /var/lib/garb
ExecStartPre=/usr/bin/chmod go-rwx        /var/lib/garb
```
/etc/systemd/system/garb.service.d/02-logfile.conf:
```
[Service]
ExecStartPre=/usr/bin/touch               /var/log/garb.log
ExecStartPre=/usr/bin/chown nobody:nobody /var/log/garb.log
ExecStartPre=/usr/bin/chmod go-rwx        /var/log/garb.log
```
/usr/bin/garb-systemd, adding chdir right before starting the acual daemon:
```
program_start() {
        echo "Starting garbd"
        if [ -d "${WORKING_DIRECTORY}" ]; then cd "${WORKING_DIRECTORY}"; fi
        /usr/bin/garbd "$@"
}
```
/etc/sysconfig/garb: dd variable WORKING_DIRECTORY
```
WORKING_DIRECTORY=/var/lib/garb
```
( Edited multiple times :)

**Comment by csuka** [comment]
Still doesn't work with the changes you supplied @ LudovicEnault
Using CentOS 8.

-- edit --

seems that the log file is garbd.log, instead of garb.log... then it works

**Comment by emcepe** [comment]
Its 2021 now, newest garbd installed and the bug is still there.

Come on, can we please! get a fixed version w/o all the hackeritis in init scripts which all do not! work at all because its hardcoded in the binary!

Thanks.

**Comment by aduzsardi** [comment]
I just replaced the init script with a systemd service

```
# rm -rf /etc/init.d/garb
# rm -rf /run/systemd/generator.late/garb.service


# cat > /etc/systemd/system/garb.service <<-'EOF'
[Unit]
Description=Galera Arbitrator 3
After=network.target

[Service]
EnvironmentFile=-/etc/default/garb
ExecStartPre=/usr/bin/chown -R nobody:nogroup /var/lib/garb
ExecStartPre=/bin/bash -c "if [[ -z $GALERA_NODES ]]; then echo 'GALERA_NODES variable not set in /etc/default/garb'; exit 1; fi"
ExecStartPre=/bin/bash -c "if [[ -z $GALERA_GROUP ]]; then echo 'GALERA_GROUP variable not set /etc/default/garb'; exit 1; fi"
ExecStart=/usr/bin/garbd -n $GALERA_NODE_NAME -a gcomm://${GALERA_NODES} -g $GALERA_GROUP -o $GALERA_OPTIONS -l $LOG_FILE
WorkingDirectory=/var/lib/garb
User=nobody
Group=nogroup
PrivateTmp=false
PermissionsStartOnly=true

[Install]
WantedBy=multi-user.target
EOF


# systemctl daemon-reload && systemctl enable garb && systemctl start garb
```

**Note: the /var/lib/garb directory needs to be created beforehand**

**Comment by Habeeb556** [comment]
This issue still exists in version `4.20r`. The service is being created with the user set to `nobody` (configured with `User=nobody` in the service file), which has limited permissions. To resolve this, you need to create a dedicated user for the service. For example, you can use the following command:

```bash
useradd --system --no-create-home --shell /sbin/nologin garb
```

Next, create the required directory:

```bash
mkdir -p /var/lib/galera
```

Then, add the following line to `/etc/sysconfig/garb` to configure the base directory for Galera:

```bash
GALERA_OPTIONS="base_dir=/var/lib/galera"
```

Finally, ensure the proper ownership and permissions for the directory:

```bash
chown garb:garb /var/lib/galera
```

After setting the permissions, restart the service to apply the changes.
